drivers: pinctrl: Add ZynqMP / Mercury XU pinctrl support

jbylicki opened the pull request:
This PR removes the custom pin control setup from the `board.c` file for the Mercury XU platform and replaces it with an implementation compliant with the `pinctrl` specification.

Fixes #52437

jbylicki committed:
drivers: pinctrl: Add ZynqMP / Mercury XU pinctrl support

Add a pinctrl driver for the ZynqMP SoC and the
Mercury XU board powered by it.

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki committed:
boards: arm: mercury_xu: Add remaining UART pin definitions

Add remaining UART0 pin definitions and migrate UART1 to utilize the
pinmux api

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

gmarull commented on drivers/pinctrl/pinctrl_xlnx_zynqmp.c:
```diff
@@ -0,0 +1,42 @@
+
+#define DT_DRV_COMPAT xlnx_pinctrl_zynqmp
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, "Unsupported number of instances");
```
redundant

gmarull commented on drivers/pinctrl/pinctrl_xlnx_zynqmp.c:
```diff
@@ -0,0 +1,42 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "pinctrl_soc.h"
+#include <zephyr/dt-bindings/pinctrl/pinctrl-zynqmp.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/syscon.h>
+#include <zephyr/devicetree.h>
```
please, cleanup includes. not everything looks necessary

gmarull commented on drivers/pinctrl/pinctrl_xlnx_zynqmp.c:
```diff
@@ -0,0 +1,42 @@
+
+LOG_MODULE_REGISTER(pinctrl_xlnx_zynqmp, CONFIG_PINCTRL_LOG_LEVEL);
+
+#define DT_DRV_COMPAT xlnx_pinctrl_zynqmp
```
driver does not use any DT APIs, other than the redundant line below.

jbylicki replied:
Now it does use DT API to fetch the MIO Register base address. Should it be removed despite that?

gmarull commented on dts/bindings/pinctrl/xlnx,pinctrl-zynqmp.yaml:
```diff
@@ -0,0 +1,28 @@
+    Definitions for a pinctrl state.
+  child-binding:
+
+    include:
+      - name: pincfg-node.yaml
+        property-allowlist:
```
if not allowing anything, just skip this

gmarull commented on include/zephyr/dt-bindings/pinctrl/pinctrl-zynqmp.h:
```diff
@@ -0,0 +1,16 @@
+#define UART_FUNCTION 0x1
+
+#define UART0_RX_38 (38U | (UART_FUNCTION << 8))
+#define UART0_TX_39 (39U | (UART_FUNCTION << 8))
```
define magic 8?

gmarull commented on soc/arm/xilinx_zynqmp/pinctrl_soc.h:
```diff
@@ -0,0 +1,58 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/types.h>
+
+#define MIO_DEFAULT	   0x0
+#define MIO_BASE_ADDR  0xff180000
+#define MIO_PIN_OFFSET 0x04
```
please no addresses out of DT, is this even needed?

gmarull commented on soc/arm/xilinx_zynqmp/pinctrl_soc.h:
```diff
@@ -0,0 +1,58 @@
+#define MIO_BASE_ADDR  0xff180000
+#define MIO_PIN_OFFSET 0x04
+
+#define MIO_L0_SEL     BIT(1)
+#define MIO_L1_SEL     BIT(2)
```
not used?

jbylicki replied:
Yes, that's unused, but I think it should be left here. It defines the layout of the register. One less thing to check for the next person using it.

gmarull replied:
dead code is not allowed in Zephyr

henrikbrixandersen replied:
I don't think this qualifies as dead code. A lot of device drivers and SoCs define all of their registers and fields, even if they are not used yet.

fkokosinski replied:
Hi @henrikbrixandersen, @gmarull! It looks like the discussion here has been stalled for some time now.

We’ve originally added the `MIO_L1_SEL` field definition for reasons stated by @henrikbrixandersen, but we’re okay with either way - keeping this define or removing it altogether.

Let us know how we should proceed with this. Thanks!

henrikbrixandersen replied:
I'd keep it.

gmarull replied:
Ok, it won't hurt, it's just a single definition with potential future use within pinctrl_soc.h.

gmarull reviewed (changes requested):
please split board changes from driver commit

jbylicki committed:
boards: arm: mercury_xu: Define the pinctrl instance

Define the pinctrl-based pin controller instance
for the Mercury XU board and remove the old implementation

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>
